File: packages/asdl-pr-address/src/asdl_pr_address/cli/pr_address/stack_feedback_thread_index.py
"""Canonical stack feedback review-thread identity helpers."""

from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass
from typing import TypeAlias

from asdl_pr_address.cli.pr_address.stack_feedback import (
    StackFeedbackPlanBatch,
    StackFeedbackPlanInformationalItem,
    StackFeedbackPlanItem,
    StackFeedbackPlanResult,
)
from asdl_pr_address.cli.pr_address.string_values import trim_optional

ThreadKey: TypeAlias = tuple[int, str]
StackFeedbackPlanThreadItem: TypeAlias = StackFeedbackPlanItem | StackFeedbackPlanInformationalItem


@dataclass(frozen=True)
class StackFeedbackThreadIndex:
    planned_pr_numbers: tuple[int, ...]
    actionable_review_threads: tuple[StackFeedbackPlanItem, ...]
    actionable_review_thread_keys: set[ThreadKey]
    known_review_thread_keys: set[ThreadKey]
    informational_review_thread_keys: set[ThreadKey]


def build_stack_feedback_thread_index(
    stack_plan: StackFeedbackPlanResult,
) -> StackFeedbackThreadIndex:
    actionable_items = actionable_review_thread_items(stack_plan)
    actionable_keys = _review_thread_keys(actionable_items)
    informational_keys = informational_review_thread_keys(stack_plan)
    return StackFeedbackThreadIndex(
        planned_pr_numbers=planned_pr_numbers(stack_plan),
        actionable_review_threads=actionable_items,
        actionable_review_thread_keys=actionable_keys,
        known_review_thread_keys=actionable_keys | informational_keys,
        informational_review_thread_keys=informational_keys,
    )


def thread_key(pr_number: int, thread_id: str | None) -> ThreadKey | None:
    trimmed = trim_optional(thread_id)
    if trimmed is None:
        return None
    return (pr_number, trimmed)


def planned_pr_numbers(stack_plan: StackFeedbackPlanResult) -> tuple[int, ...]:
    if stack_plan.validation.per_pr:
        return tuple(item.pr_number for item in stack_plan.validation.per_pr)

    pr_numbers: list[int] = []
    for batch in stack_plan.batches:
        for item in batch.items:
            pr_numbers.append(item.pr_number)
    for item in stack_plan.informational:
        pr_numbers.append(item.pr_number)
    return tuple(dict.fromkeys(pr_numbers))


def actionable_review_thread_items(
    stack_plan: StackFeedbackPlanResult,
) -> tuple[StackFeedbackPlanItem, ...]:
    return tuple(
        item
        for batch in stack_plan.batches
        for item in batch.items
        if item.source_kind == "review_thread"
    )


def selected_batch_review_thread_items(
    selected_batch: StackFeedbackPlanBatch,
) -> tuple[StackFeedbackPlanItem, ...]:
    return tuple(item for item in selected_batch.items if item.source_kind == "review_thread")


def known_review_thread_keys(stack_plan: StackFeedbackPlanResult) -> set[ThreadKey]:
    actionable_keys = _review_thread_keys(actionable_review_thread_items(stack_plan))
    return actionable_keys | informational_review_thread_keys(stack_plan)


def informational_review_thread_keys(stack_plan: StackFeedbackPlanResult) -> set[ThreadKey]:
    return _review_thread_keys(
        tuple(item for item in stack_plan.informational if item.source_kind == "review_thread")
    )


def items_by_thread(
    items: tuple[StackFeedbackPlanItem, ...],
) -> dict[str, tuple[StackFeedbackPlanItem, ...]]:
    grouped: dict[str, list[StackFeedbackPlanItem]] = defaultdict(list)
    for item in items:
        trimmed_thread_id = trim_optional(item.thread_id)
        if trimmed_thread_id is not None:
            grouped[trimmed_thread_id].append(item)
    return {thread_id: tuple(thread_items) for thread_id, thread_items in grouped.items()}


def other_batch_review_threads(
    *,
    plan: StackFeedbackPlanResult,
    selected_batch_id: str,
) -> dict[ThreadKey, str]:
    lookup: dict[ThreadKey, str] = {}
    for batch in plan.batches:
        if batch.batch_id == selected_batch_id:
            continue
        for item in batch.items:
            if item.source_kind != "review_thread":
                continue
            key = thread_key(item.pr_number, item.thread_id)
            if key is not None:
                lookup[key] = batch.batch_id
    return lookup


def duplicate_thread_keys(keys: tuple[ThreadKey, ...]) -> tuple[ThreadKey, ...]:
    seen: set[ThreadKey] = set()
    duplicates: list[ThreadKey] = []
    for key in keys:
        if key in seen and key not in duplicates:
            duplicates.append(key)
        seen.add(key)
    return tuple(duplicates)


def _review_thread_keys(items: tuple[StackFeedbackPlanThreadItem, ...]) -> set[ThreadKey]:
    keys: set[ThreadKey] = set()
    for item in items:
        key = thread_key(item.pr_number, item.thread_id)
        if key is not None:
            keys.add(key)
    return keys

File: ts/packages/pr-address/src/stack-feedback-thread-index.ts

export type ThreadKey = readonly [prNumber: number, threadId: string];

export interface ThreadIndexItem {
	readonly pr_number: number;
	readonly thread_id?: string | null | undefined;
	readonly source_kind: string;
}

export interface ThreadIndexBatch<TItem extends ThreadIndexItem> {
	readonly batch_id: string;
	readonly items: readonly TItem[];
}

export interface ThreadIndexPlan<TItem extends ThreadIndexItem> {
	readonly batches: readonly ThreadIndexBatch<TItem>[];
	readonly informational: readonly TItem[];
}

export interface ThreadIndexValidation {
	readonly per_pr: readonly { readonly pr_number: number }[];
}


export function threadKey(prNumber: number, threadId: string | null | undefined): ThreadKey | null {
	const trimmed = trimOptional(threadId);
	if (trimmed === null) return null;
	return [prNumber, trimmed];
}

export function threadKeyString(prNumber: number, threadId: string): string {
	return `${prNumber}\u0000${threadId}`;
}

export function plannedPrNumbers<TItem extends ThreadIndexItem>(
	stackPlan: ThreadIndexPlan<TItem> & { readonly validation: ThreadIndexValidation },
): number[] {
	if (stackPlan.validation.per_pr.length > 0) return stackPlan.validation.per_pr.map((item) => item.pr_number);
	const prNumbers: number[] = [];
	for (const batch of stackPlan.batches) {
		for (const item of batch.items) prNumbers.push(item.pr_number);
	}
	for (const item of stackPlan.informational) prNumbers.push(item.pr_number);
	return [...new Set(prNumbers)];
}

export function actionableReviewThreadItems<TItem extends ThreadIndexItem>(stackPlan: ThreadIndexPlan<TItem>): TItem[] {
	return stackPlan.batches.flatMap((batch) => batch.items.filter((item) => item.source_kind === "review_thread"));
}


export function knownReviewThreadKeys<TItem extends ThreadIndexItem>(stackPlan: ThreadIndexPlan<TItem>): Set<string> {
	const keys = reviewThreadKeys(actionableReviewThreadItems(stackPlan));
	for (const key of informationalReviewThreadKeys(stackPlan)) keys.add(key);
	return keys;
}

export function informationalReviewThreadKeys<TItem extends ThreadIndexItem>(
	stackPlan: Pick<ThreadIndexPlan<TItem>, "informational">,
): Set<string> {
	return reviewThreadKeys(stackPlan.informational.filter((item) => item.source_kind === "review_thread"));
}

export function itemsByThread<TItem extends ThreadIndexItem>(items: readonly TItem[]): Map<string, TItem[]> {
	const grouped = new Map<string, TItem[]>();
	for (const item of items) {
		const threadId = trimOptional(item.thread_id);
		if (threadId === null) continue;
		const existing = grouped.get(threadId);
		if (existing === undefined) grouped.set(threadId, [item]);
		else existing.push(item);
	}
	return grouped;
}

export function otherBatchReviewThreads<TItem extends ThreadIndexItem>(
	stackPlan: ThreadIndexPlan<TItem>,
	selectedBatchId: string,
): Map<string, string> {
	const lookup = new Map<string, string>();
	for (const batch of stackPlan.batches) {
		if (batch.batch_id === selectedBatchId) continue;
		for (const item of batch.items) {
			if (item.source_kind !== "review_thread") continue;
			const key = threadKey(item.pr_number, item.thread_id);
			if (key !== null) lookup.set(threadKeyString(key[0], key[1]), batch.batch_id);
		}
	}
	return lookup;
}

export function duplicateThreadKeys(keys: readonly ThreadKey[]): ThreadKey[] {
	const seen = new Set<string>();
	const duplicates: ThreadKey[] = [];
	const duplicateStrings = new Set<string>();
	for (const key of keys) {
		const serialized = threadKeyString(key[0], key[1]);
		if (seen.has(serialized) && !duplicateStrings.has(serialized)) {
			duplicates.push(key);
			duplicateStrings.add(serialized);
		}
		seen.add(serialized);
	}
	return duplicates;
}

function reviewThreadKeys(items: readonly ThreadIndexItem[]): Set<string> {
	const keys = new Set<string>();
	for (const item of items) {
		const key = threadKey(item.pr_number, item.thread_id);
		if (key !== null) keys.add(threadKeyString(key[0], key[1]));
	}
	return keys;
}


function trimOptional(value: string | null | undefined): string | null {
	if (value === null || value === undefined) return null;
	const trimmed = value.trim();
	return trimmed === "" ? null : trimmed;
}
